runtime/src/main/java/org/corfudb/runtime/Restore.java

```diff
@@ -227,8 +234,71 @@ private void clearAllTables() {
         txn.commit();
     }
 
+    private void nonCachedAppendSMREntries(UUID streamId, SMREntry... smrEntries) {
+        MultiSMREntry multiSMREntry = new MultiSMREntry();
+        multiSMREntry.addTo(Arrays.asList(smrEntries));
+        rt.getStreamsView().append(multiSMREntry, null, CacheOption.WRITE_AROUND, streamId);
+    }
+
     public enum RestoreMode {
         FULL,   // Clean up ALL tables before restore
         PARTIAL // Clean up ONLY tables that are to be restored
     }
+
+    class StreamBatchWriter {
+
+        private final int maxBatchSize;
+        private final int maxWriteSize;
+        private final UUID streamId;
+
+        @Getter
+        private int totalNumSMREntries = 0;
+        @Getter
+        private int totalWriteSize = 0;
+
+        private final List<SMREntry> buffer = new ArrayList<>();
+        private int bufferWriteSize = 0;
+
+        StreamBatchWriter(int maxBatchSize, int maxWriteSize, UUID streamId) {
+            this.maxBatchSize = maxBatchSize;
+            this.maxWriteSize = maxWriteSize;
+            this.streamId = streamId;
+        }
+
+        public void batchWrite(List<SMREntry> smrEntries) {
+            if (smrEntries.isEmpty()) {
+                return;
+            }
+
+            Preconditions.checkState(buffer.size() < maxBatchSize);
+
+            for (SMREntry smrEntry : smrEntries) {
+                if (buffer.size() == maxBatchSize || bufferWriteSize + smrEntry.getSerializedSize() > maxWriteSize) {
+                    flushBuffer();
+                }
+                buffer.add(smrEntry);
+                bufferWriteSize += smrEntry.getSerializedSize();
+            }
+
+            // Keep buffer not full
+            if (buffer.size() == maxBatchSize || bufferWriteSize > maxWriteSize) {
```

Comment:
Why is this condition needed. If by the end of the for-loop the buffer is full, then there are two cases:

1. A new write will cause the buffer to be flushed
2. No new write, and the buffer will be flushed by `shutdown` which flushes the buffer

Is redundant, or am I missing another case where the buffer isn't covered by those two cases and won't be written out?

Reply:
It is redundant in terms of correctness. I was originally thinking to have batchWrite() flushing buffer as soon as it gets full. For simplicity it is better to remove this condition.